engine/apps/alerts/tasks/notify_user.py

```diff
@@ -170,51 +194,97 @@ def notify_user_task(
                         notification_error_code=UserNotificationPolicyLogRecord.ERROR_NOTIFICATION_POSTING_TO_SLACK_IS_DISABLED,
                     )
                 else:
-                    log_record = UserNotificationPolicyLogRecord(
-                        author=user,
-                        type=UserNotificationPolicyLogRecord.TYPE_PERSONAL_NOTIFICATION_TRIGGERED,
-                        notification_policy=notification_policy,
-                        alert_group=alert_group,
-                        reason=reason,
-                        slack_prevent_posting=prevent_posting_to_thread,
-                        notification_step=notification_policy.step,
-                        notification_channel=notification_policy.notify_by,
+                    # todo: feature flag
+                    user_notification_bundle, _ = UserNotificationBundle.objects.get_or_create(
+                        user=user, important=important, notification_channel=notification_policy.notify_by
                     )
-        if log_record:  # log_record is None if user notification policy step is unspecified
-            # if this is the first notification step, and user hasn't been notified for this alert group - update metric
-            if (
-                log_record.type == UserNotificationPolicyLogRecord.TYPE_PERSONAL_NOTIFICATION_TRIGGERED
-                and previous_notification_policy_pk is None
-                and not user.personal_log_records.filter(
-                    type=UserNotificationPolicyLogRecord.TYPE_PERSONAL_NOTIFICATION_TRIGGERED,
-                    alert_group_id=alert_group_pk,
-                ).exists()
-            ):
-                update_metrics_for_user.apply_async((user.id,))
+                    user_notification_bundle = UserNotificationBundle.objects.filter(
+                        pk=user_notification_bundle.pk
+                    ).select_for_update()[0]
+                    # check if notification needs to be bundled
+                    if (
+                        notification_policy.notify_by in UserNotificationBundle.NOTIFICATION_CHANNELS_TO_BUNDLE
+                        and user_notification_bundle.notified_recently()
+                    ):
+                        user_notification_bundle.notification_data.append(
```

Comment:
We might want some upper limit on what can be appended here.  It's an edge case but at some point the bundle has so much in it that it's not useful it may save resources to just discard new entries at beyond that point.